linkerd/stack/src/switch.rs

```diff
@@ -0,0 +1,84 @@
+//! A middleware that switches between two underlying stacks, depending on the
+//! target type.
+
+use futures::{future, prelude::*};
+use linkerd2_error::Error;
+use std::task::{Context, Poll};
+use tower::util::ServiceExt;
+
+/// Determines whether the primary stack should be used.
+pub trait Switch<T> {
+    fn use_primary(&self, target: &T) -> bool;
+}
+
+/// Makes either the primary or fallback stack, as determined by an `S`-typed
+/// `Switch`.
+pub struct MakeSwitch<S, P, F> {
```

Comment:
nit: not sold on calling this `MakeSwitch`, since it would work equally well at the `Service`/`Request` level as well as the `MakeService`/`Target` level...names starting with `Make` kinda suggest to me that this only ever has a `Service` response type. Maybe the service should be `Switch` and the trait `Predicate or something?